be/ut/alloc.c

```diff
@@ -398,6 +398,204 @@ M0_INTERNAL void m0_be_ut_alloc_spare(void)
+				 m0_be_alloc_stats_capture(a, tx)));
+			ut_ptr[i] = NULL;
+		}
+	}
+
+	/**
+	 * 1. Allocate multiple chunks with some having chunk_align set to true
+	 *    and the remaining having chunk_align as false.
+	 * 2. Verify the alignment of all the allocated chunks.
+	 * 3. Delete some of the allocated chunks.
+	 * 4. Verify the alignment of the remaining chunks.
+	 * 5. Delete remaining of the allocated chunks.
+	 */
+
+	M0_SET_ARR0(ut_ptr);
+
+	/**
+	 *  Alloc some chunks with chunk_align parameter set as true and
+	 *  remaining with chunk_align set as false.
+	 */
+	for (i = 0; i < ut_nr; i++) {
+		j = i;
+		ut_size  = m0_rnd64(&j) % BE_UT_ALLOC_SIZE + 1;
+		ut_tval  = i % 2 == 0 ? true : false;
+		ut_shift = i % 2 == 0 ? BE_UT_ALLOC_SHIFT - 1 :
+					BE_UT_ALLOC_SHIFT;
+
+		M0_BE_UT_TRANSACT(ut_be, tx, cred,
+			  (m0_be_allocator_credit(a, M0_BAO_ALLOC_ALIGNED,
+						  ut_size, ut_shift, &cred),
+			   m0_be_alloc_stats_credit(a, &cred)),
+			  (M0_BE_OP_SYNC(op,
+					 m0_be_alloc_aligned(a, tx, &op,
+						     &ut_ptr[i],
+						     ut_size, ut_shift,
+						     M0_BITS(M0_BAP_NORMAL),
+						     ut_tval)),
+			   m0_be_alloc_stats_capture(a, tx)));
+		M0_UT_ASSERT(ut_ptr[i] != NULL);
+	}
+
+	/* Verify the alignment of the chunks */
+	for (i = 0; i < ut_nr; i++) {
+		if (i%2 == 0)
+			M0_UT_ASSERT(m0_addr_is_aligned(ut_ptr[i],
+							BE_UT_ALLOC_SHIFT - 1));
+		else
+			M0_UT_ASSERT(m0_addr_is_aligned(ut_ptr[i],
+							BE_UT_ALLOC_SHIFT));
+	}
+
+	/**
+	 *  Delete every third chunk to make sure that both type of chunks are
+	 *  deleted.
+	 */
+	for (i = 0; i < ut_nr; i += 3) {
+		j = i;
+		ut_size  = m0_rnd64(&j) % BE_UT_ALLOC_SIZE + 1;
+		ut_shift = i % 2 == 0 ? BE_UT_ALLOC_SHIFT - 1 :
+					BE_UT_ALLOC_SHIFT;
+		M0_BE_UT_TRANSACT(ut_be, tx, cred,
+			  (m0_be_allocator_credit(a, M0_BAO_FREE_ALIGNED,
+						  ut_size, ut_shift, &cred),
+			   m0_be_alloc_stats_credit(a, &cred)),
+			  (M0_BE_OP_SYNC(op,
+					 m0_be_free_aligned(a, tx, &op,
+							    ut_ptr[i])),
+			   m0_be_alloc_stats_capture(a, tx)));
+		ut_ptr[i] = NULL;
+	}
+
+	/* Verify the alignment of the remaining chunks */
+	for (i = 0; i < ut_nr; i++) {
+		if (ut_ptr[i] != NULL) {
+			if (i%2 == 0)
+				M0_UT_ASSERT(m0_addr_is_aligned(ut_ptr[i],
+							BE_UT_ALLOC_SHIFT - 1));
+			else
+				M0_UT_ASSERT(m0_addr_is_aligned(ut_ptr[i],
+							    BE_UT_ALLOC_SHIFT));
+		}
+	}
+
+	/* Delete remaining chunks */
+	for (i = 0; i < ut_nr; i++) {
+		if (ut_ptr[i] != NULL) {
+			j = i;
+			ut_size  = m0_rnd64(&j) % BE_UT_ALLOC_SIZE + 1;
+			ut_shift = i % 2 == 0 ? BE_UT_ALLOC_SHIFT - 1 :
+						BE_UT_ALLOC_SHIFT;
+			M0_BE_UT_TRANSACT(ut_be, tx, cred,
+				(m0_be_allocator_credit(a, M0_BAO_FREE_ALIGNED,
+							ut_size, ut_shift,
+							&cred),
+				 m0_be_alloc_stats_credit(a, &cred)),
+				(M0_BE_OP_SYNC(op,
+					       m0_be_free_aligned(a, tx, &op,
+								  ut_ptr[i])),
+				 m0_be_alloc_stats_capture(a, tx)));
+			ut_ptr[i] = NULL;
+		}
+	}
+
```

Comment:
There is a function m0_be_allocator__invariant() in alloc.c. You can use it to make sure that everything is fine with the allocated memory.

Reply:
It is called by m0_be_alloc_aligned only when expensive checks are turned ON. Confirm if this is true when ut's are run and if-not then explicitly make a call to the invariant in your ut.

Reply:
Added calls for m0_be_allocator__invariant at the start of the UT, after allocation of memory, after freeing up of memory and at the end of the UT.